internals/overlord/state/identities.go

```diff
@@ -0,0 +1,288 @@
+			missing = append(missing, name)
+		}
+		err := identity.validate()
+		if err != nil {
+			return fmt.Errorf("identity %q invalid: %w", name, err)
+		}
+	}
+	if len(missing) > 0 {
+		sort.Strings(missing)
+		return fmt.Errorf("identities do not exist: %s", strings.Join(missing, ", "))
+	}
+
+	newIdentities := s.cloneIdentities()
+	for name, identity := range identities {
+		identity.Name = name
+		newIdentities[name] = identity
+	}
+
+	err := verifyUniqueUserIDs(newIdentities)
+	if err != nil {
+		return err
+	}
+
+	s.identities = newIdentities
+	return nil
+}
+
+// ReplaceIdentities replaces the named identities in the system with the
+// given identities (adding those that don't exist), or removes them if the
+// map value is nil.
+func (s *State) ReplaceIdentities(identities map[string]*Identity) error {
+	s.writing()
+
+	for name, identity := range identities {
+		if identity != nil {
+			err := identity.validate()
+			if err != nil {
+				return fmt.Errorf("identity %q invalid: %w", name, err)
+			}
+		}
+	}
+
+	newIdentities := s.cloneIdentities()
+	for name, identity := range identities {
+		if identity == nil {
+			delete(newIdentities, name)
+		} else {
+			identity.Name = name
+			newIdentities[name] = identity
+		}
+	}
+
+	err := verifyUniqueUserIDs(newIdentities)
+	if err != nil {
+		return err
+	}
+
+	s.identities = newIdentities
+	return nil
+}
+
+// RemoveIdentities removes the named identities from the system. It's an
+// error if any of the named identities do not exist.
+func (s *State) RemoveIdentities(identities map[string]struct{}) error {
+	s.writing()
+
+	// If any of the named identities don't exist, return an error.
+	var missing []string
+	for name := range identities {
+		if _, ok := s.identities[name]; !ok {
+			missing = append(missing, name)
+		}
+	}
+	if len(missing) > 0 {
+		sort.Strings(missing)
+		return fmt.Errorf("identities do not exist: %s", strings.Join(missing, ", "))
+	}
+
+	for name := range identities {
+		delete(s.identities, name)
+	}
+	return nil
+}
+
+// Identities returns all the identities in the system. The returned map is a
+// shallow clone, so map mutations won't affect state.
+func (s *State) Identities() map[string]*Identity {
+	s.reading()
+
+	result := make(map[string]*Identity, len(s.identities))
+	for name, identity := range s.identities {
+		result[name] = identity
+	}
+	return result
+}
+
+func (s *State) cloneIdentities() map[string]*Identity {
+	newIdentities := make(map[string]*Identity, len(s.identities))
+	for name, identity := range s.identities {
+		newIdentities[name] = identity
+	}
+	return newIdentities
+}
+
+func verifyUniqueUserIDs(identities map[string]*Identity) error {
+	userIDs := make(map[uint32]string) // maps user IDs to identity name
+	for name, identity := range identities {
+		switch {
+		case identity.Local != nil:
+			existingName, ok := userIDs[identity.Local.UserID]
+			if ok {
+				if name > existingName { // ensure error message is stable
```

Comment:
This makes the error message stable between two identities. But if there is 3 or more identities with the same UID, then this does not work due to go map pseudo random iteration order.

Reply:
Good call, updated the function and error message.